File: logger/src/main/java/fr/beapp/logger/formatter/DefaultFormatter.java
package fr.beapp.logger.formatter;


import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Locale;
import java.util.WeakHashMap;

import fr.beapp.logger.Logger;

public class DefaultFormatter implements Formatter {

	private static final WeakHashMap<String, String> SHORTEN_CLASS_NAMES = new WeakHashMap<>();

	private final String stacktraceSentinel;

	public DefaultFormatter() {
		this(Logger.class.getCanonicalName());
	}

	public DefaultFormatter(String stacktraceSentinel) {
		this.stacktraceSentinel = stacktraceSentinel;
	}

	@Nullable
	@Override
	public String format(@Nullable Throwable tr, @Nullable String message, Object... args) {
		if (message == null || message.isEmpty()) {
			message = null;
		}

		if (message == null) {
			if (tr == null) {
				return null; // Swallow message if it's null and there's no throwable.
			}
			message = getStackTraceString(tr);
		} else {
			if (args.length > 0) {
				message = String.format(message, args);
			}
			if (tr != null) {
				message += "\n" + getStackTraceString(tr);
			}
		}

		String caller = retrieveCaller();
		if (caller == null)
			return message;

		return String.format("%s - %s", caller, message);
	}

	@Nullable
	protected String getStackTraceString(@NonNull Throwable t) {
		// Don't replace this with Log.getStackTraceString() - it hides UnknownHostException, which is not what we want.
		StringWriter sw = new StringWriter(256);
		PrintWriter pw = new PrintWriter(sw, false);
		try {
			t.printStackTrace(pw);
			pw.flush();
			return sw.toString();
		} finally {
			pw.close();
		}
	}

	@Nullable
	protected String retrieveCaller() {
		try {
			StackTraceElement[] stackTraceElements = Thread.currentThread().getStackTrace();

			boolean isInLoggerClass = false;
			for (StackTraceElement stackTraceElement : stackTraceElements) {
				if (stacktraceSentinel.equals(stackTraceElement.getClassName())) {
					isInLoggerClass = true;
				} else if (isInLoggerClass) {
					return String.format(Locale.ENGLISH, "%s:%d", shortenCanonicalName(stackTraceElement.getClassName()), stackTraceElement.getLineNumber());
				}
			}
		} catch (Exception ignored) {
			// Can't use Logger.error here
		}
		return null;
	}

	@NonNull
	protected String shortenCanonicalName(@NonNull final String canonicalName) {
		if (canonicalName.isEmpty())
			return "";

		synchronized (SHORTEN_CLASS_NAMES) {
			String shortenClassName = SHORTEN_CLASS_NAMES.get(canonicalName);
			if (shortenClassName == null) {
				StringBuilder shortenName = new StringBuilder();

				int lastIndex = 0;
				int nextSep = -1;
				do {
					shortenName.append(canonicalName.charAt(nextSep + 1));
					nextSep = canonicalName.indexOf('.', nextSep + 1);

					if (nextSep > -1) {
						shortenName.append('.');
						lastIndex = nextSep + 1;
					}
				} while (nextSep > -1);

				if (lastIndex < canonicalName.length() - 1) {
					shortenName.append(canonicalName.substring(lastIndex + 1));
				}

				shortenClassName = shortenName.toString();
				SHORTEN_CLASS_NAMES.put(canonicalName, shortenClassName);
			}

			return shortenClassName;
		}
	}

}

File: logger/src/main/java/fr/beapp/logger/formatter/DefaultFormatter.kt
package fr.beapp.logger.formatter

import fr.beapp.logger.Logger
import java.io.PrintWriter
import java.io.StringWriter
import java.util.Locale
import java.util.WeakHashMap

public open class DefaultFormatter(private val stacktraceSentinel: String? = Logger::class.java.canonicalName) : Formatter {
	private companion object {
		val SHORTEN_CLASS_NAMES = WeakHashMap<String, String>()
	}

	override fun format(tr: Throwable?, message: String?, vararg args: Any?): String? {
		val computedMessage = when {
			// Swallow message if it's null and there's no throwable.
			message == null -> tr?.let(::getStackTraceString) ?: return null
			else -> {
				// format arguments if there's any
				val formattedMessage = if (args.isEmpty()) message
				else String.format(message, *args)
				// include stacktrace if there's a throwable
				if (tr != null) "$formattedMessage\n${getStackTraceString(tr)}"
				else formattedMessage
			}
		}
		val caller = retrieveCaller() ?: return computedMessage
		return String.format("%s - %s", caller, computedMessage)
	}

	public open fun getStackTraceString(t: Throwable): String? {
		// Don't replace this with Log.getStackTraceString() - it hides UnknownHostException, which is not what we want.
		val sw = StringWriter(256)
		val pw = PrintWriter(sw, false)
		return pw.use { printWriter ->
			t.printStackTrace(printWriter)
			printWriter.flush()
			sw.toString()
		}
	}

	public open fun retrieveCaller(): String? {
		try {
			val stackTraceElements = Thread.currentThread().stackTrace
			var isInLoggerClass = false
			for (stackTraceElement in stackTraceElements) {
				if (stacktraceSentinel == stackTraceElement.className) {
					isInLoggerClass = true
				} else if (isInLoggerClass) {
					return String.format(Locale.ENGLISH, "%s:%d", shortenCanonicalName(stackTraceElement.className), stackTraceElement.lineNumber)
				}
			}
		} catch (ignored: Exception) {
			// Can't use Logger.error here
		}
		return null
	}

	public open fun shortenCanonicalName(canonicalName: String): String {
		if (canonicalName.isEmpty()) return ""
		synchronized(SHORTEN_CLASS_NAMES) {
			var shortenClassName = SHORTEN_CLASS_NAMES[canonicalName]
			if (shortenClassName == null) {
				val shortenName = StringBuilder()
				var lastIndex = 0
				var nextSep = -1
				do {
					shortenName.append(canonicalName[nextSep + 1])
					nextSep = canonicalName.indexOf('.', nextSep + 1)
					if (nextSep > -1) {
						shortenName.append('.')
						lastIndex = nextSep + 1
					}
				} while (nextSep > -1)
				if (lastIndex < canonicalName.length - 1) {
					shortenName.append(canonicalName.substring(lastIndex + 1))
				}
				shortenClassName = shortenName.toString()
				SHORTEN_CLASS_NAMES[canonicalName] = shortenClassName
			}
			return shortenClassName
		}
	}


}